### PR TITLE
Broker requests

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#import "MSIDBrokerOperationInteractiveTokenRequest.h"
+#import "MSIDInteractiveRequestParameters.h"
+
+@implementation MSIDBrokerOperationInteractiveTokenRequest (InteractiveParameters)
+
++ (instancetype)tokenRequestWithParameteres:(MSIDInteractiveRequestParameters *)parameters
+                                      error:(NSError **)error
+{
+    __auto_type request = [MSIDBrokerOperationInteractiveTokenRequest new];
+    BOOL result = [self fillRequest:request withParameteres:parameters error:error];
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m]: nit: typo in "withParameteres"

File: IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.m [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <AuthenticationServices/AuthenticationServices.h>
+#import "MSIDBrokerOperationSilentTokenRequest+SSORequest.h"
+#import "MSIDJsonSerializer.h"
+
+@implementation MSIDBrokerOperationSilentTokenRequest (SSORequest)
+
+- (ASAuthorizationSingleSignOnRequest *)ssoRequestWithProvider:(ASAuthorizationSingleSignOnProvider *)provider
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.m]: would this code be different or same between silent and interactive?

File: IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    if (!result) return nil;
+    
+    request.accountIdentifier = parameters.accountIdentifier;
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m]: would this also need claims and client capabilities, or those are just basic parameters and others will be added later?

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m [hunk 4]
```diff
@@ -52,18 +53,45 @@ - (void)acquireToken:(MSIDRequestCompletionBlock)completionBlock
...
-    // TODO: implement.
-    return YES;
+    return [[ASAuthorizationSingleSignOnProvider msidSharedProvider] canPerformAuthorization];
+}
+
+#pragma mark - Private
+
+- (BOOL)shouldFallback:(NSError *)error
+{
+    if (!self.fallbackController) return NO;
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: nit: add some logging here

File: IdentityCore/src/requests/MSIDInteractiveTokenRequesting.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+@class MSIDOauth2Factory;
+@class MSIDTokenResult;
+@class MSIDWebWPJResponse;
+@protocol MSIDCacheAccessor;
+
+typedef void (^MSIDInteractiveRequestCompletionBlock)(MSIDTokenResult * _Nullable result, NSError * _Nullable error, MSIDWebWPJResponse * _Nullable installBrokerResponse);
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol MSIDInteractiveTokenRequesting <NSObject>
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDInteractiveTokenRequesting.h]: why do we need a protocol for this one?

File: IdentityCore/src/requests/broker/MSIDBrokerKeyProvider.m [hunk 6]
```diff
@@ -136,6 +136,36 @@ - (NSData *)brokerKeyWithError:(NSError **)error
     return [self createBrokerKeyWithError:error];
 }
 
+- (NSString *)base64BrokerKeyWithContext:(id<MSIDRequestContext>)context
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDBrokerKeyProvider.m]: should we use this in other places for broker key as well (legacy broker)?

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    
+    if (!completionBlock)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, self.requestParameters, @"Passed nil completionBlock. End silent broker flow.");
+        return;
+    }
+    
+    NSString *upn = self.requestParameters.accountIdentifier.displayableId ?: self.requestParameters.loginHint;
+    
+    [self.requestParameters.authority resolveAndValidate:self.requestParameters.validateAuthority
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: shouldn't this be common in both interactive request cases?

File: IdentityCore/src/requests/broker/MSIDSSOExtensionSilentTokenRequest.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+// THE SOFTWARE.
+
+#import "MSIDSilentTokenRequest.h"
+
+@class MSIDAccountMetadataCacheAccessor;
+
+NS_ASSUME_NONNULL_BEGIN
+
+API_AVAILABLE(ios(13.0))
+@interface MSIDSSOExtensionSilentTokenRequest : MSIDSilentTokenRequest
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDSSOExtensionSilentTokenRequest.h]: should we subclass default silent token request instead to maximize code reuse?

File: IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+@implementation ASAuthorizationSingleSignOnProvider (MSIDExtensions)
+
++ (ASAuthorizationSingleSignOnProvider *)msidSharedProvider
+{
+    static dispatch_once_t once;
+    static ASAuthorizationSingleSignOnProvider *ssoProvider;
+    
+    dispatch_once(&once, ^{
+        // TODO: replace url.
+        NSURL *url = [NSURL URLWithString:@"https://ios-sso-test.azurewebsites.net"];
```
Comment (oldalton) [inline IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m]: this can be replaced already, right?

File: IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m]: done

File: IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.m (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.m]: Removed this category in [this pr.](https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/588). The code was almost same.

File: IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.m]: Just basic parameters. I will add other parameters later.

File: IdentityCore/src/requests/broker/MSIDBrokerKeyProvider.m (hunk 6 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDBrokerKeyProvider.m]: done

File: IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m (hunk 9 above)
Comment (antrix1989) [reply IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m]: done

File: IdentityCore/src/requests/broker/MSIDSSOExtensionSilentTokenRequest.h (hunk 8 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDSSOExtensionSilentTokenRequest.h]: There isn't much similar code between them.

File: IdentityCore/src/requests/MSIDInteractiveTokenRequesting.h (hunk 5 above)
Comment (antrix1989) [reply IdentityCore/src/requests/MSIDInteractiveTokenRequesting.h]: removed

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m (hunk 7 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: It is hard to share this logic because nested block is a bit different. Agreed to keep as it is for now.

File: IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDBrokerOperationSilentTokenRequest.h"
+#import "MSIDSSORequestProviding.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDBrokerOperationSilentTokenRequest (SSORequest) <MSIDSSORequestProviding>
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.h]: This can be part of the class itself, or is there a reason for it to be in the category?
If there's no reason, let's start from it being part of the class, and move out to categories if the first strategy doesn't work out (e.g. class becomes too big and difficult to follow)? I personally think these classes will be fairly small, so should be quite readable even without splitting.

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDBrokerOperationTokenRequest.h"
+
+@class MSIDRequestParameters;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDBrokerOperationTokenRequest (Parameteres)
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.h]: Typo: Parameter**e**s

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+@implementation MSIDBrokerOperationTokenRequest (Parameteres)
+
++ (BOOL)fillRequest:(MSIDBrokerOperationTokenRequest *)request
+     withParameters:(MSIDRequestParameters *)parameters
+              error:(NSError **)error
+{
+    NSString *accessGroup = parameters.keychainAccessGroup ?: MSIDKeychainTokenCache.defaultKeychainGroup;
+    __auto_type brokerKeyProvider = [[MSIDBrokerKeyProvider alloc] initWithGroup:accessGroup];
+        NSError *brokerError = nil;
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: nit: extra tab

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    NSString *base64UrlKey = [brokerKeyProvider base64BrokerKeyWithContext:parameters
+                                                                     error:&brokerError];
+    
+    if (!base64UrlKey)
+    {
+        // TODO: add telemetry.
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: should telemetry be common for any type of request and responsibility of the calling class?

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        if (error) *error = brokerError;
+        return NO;
+    }
+    
+    request.brokerKey = base64UrlKey;
+    request.clientVersion = [MSIDVersion sdkVersion];
+    request.protocolVersion = 4;
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: nit: can we declare shared constants for protocol versions?

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    NSDictionary *clientMetadata = parameters.appRequestMetadata;
+    request.clientAppVersion = clientMetadata[MSID_APP_VER_KEY];
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: shouldn't client version, correlationId and client app name be part of any request (not only token request) and hence be part of the super class?

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m [hunk 16]
```diff
@@ -52,18 +53,53 @@ - (void)acquireToken:(MSIDRequestCompletionBlock)completionBlock
     {
         MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Interactive broker extension flow finished. Result %@, error: %ld error domain: %@", _PII_NULLIFY(result), (long)error.code, error.domain);
         
+        if ([self shouldFallback:error])
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Falling back to local controller.");
+            
+            [self.fallbackController acquireToken:completionBlock];
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: should we return after this so that completionBlock is not executed?

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m [hunk 17]
```diff
@@ -52,18 +53,53 @@ - (void)acquireToken:(MSIDRequestCompletionBlock)completionBlock
...
+    BOOL shouldFallback = NO;
+    switch (error.code)
+    {
+        case ASAuthorizationErrorNotHandled:
+        case ASAuthorizationErrorUnknown:
+        case ASAuthorizationErrorFailed:
+            shouldFallback = YES;
+    }
+    
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"shouldFallback: %@", shouldFallback ? @"YES" : @"NO");
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: nit: provide more verbose log here, e.g. SSO extension controller should fallback : %@

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+#import "MSIDAccountIdentifier.h"
+#import "MSIDAuthority.h"
+#import "MSIDSSOExtensionTokenRequestDelegate.h"
+#import "MSIDBrokerOperationInteractiveTokenRequest+InteractiveParameters.h"
+#import "MSIDBrokerOperationInteractiveTokenRequest+SSORequest.h"
+
+@interface MSIDSSOExtensionInteractiveTokenRequest () <ASAuthorizationControllerPresentationContextProviding>
+
+@property (nonatomic) ASAuthorizationController *authorizationController;
+@property (nonatomic, copy) MSIDInteractiveRequestCompletionBlock requestCompletionBlock;
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: wouldn't base class already store the completion block? if yes, should the property be shared and exposed in internal header?

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    self = [super initWithRequestParameters:parameters
+                               oauthFactory:oauthFactory
+                     tokenResponseValidator:tokenResponseValidator
+                                 tokenCache:tokenCache
+                       accountMetadataCache:accountMetadataCache];
+
+    if (self)
+    {
+        _extensionDelegate = [MSIDSSOExtensionTokenRequestDelegate new];
+        _extensionDelegate.context = parameters;
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: nit: If I remember correctly, we generally don't store context, but pass it around. Not much difference, just consistency question.

File: IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+@implementation ASAuthorizationSingleSignOnProvider (MSIDExtensions)
+
++ (ASAuthorizationSingleSignOnProvider *)msidSharedProvider
+{
+    static dispatch_once_t once;
+    static ASAuthorizationSingleSignOnProvider *ssoProvider;
+    
+    dispatch_once(&once, ^{
+        NSURL *url = [NSURL URLWithString:@"https://login.microsoftonline.com"];
```
Comment (oldalton) [inline IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m]: nit: use constant for this one, as we should already have it declared and using in other places.

File: IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.h (hunk 10 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest+SSORequest.h]: This category was removed in this [pr](https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/588).

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.h (hunk 11 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.h]: fixed

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m (hunk 12 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: fixed

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m (hunk 15 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: These properties are defined in MSIDBrokerOperationRequest. Here we just copy them from request parameters.

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m (hunk 16 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: fixed

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m (hunk 17 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: done.

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m (hunk 18 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: we don't store it in the base class.

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m (hunk 19 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: In this case, it needs to be property because we are using it for logging in apple's callbacks, like here:

`- (void)authorizationController:(ASAuthorizationController *)controller didCompleteWithAuthorization:(ASAuthorization *)authorization`

File: IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m (hunk 20 above)
Comment (antrix1989) [reply IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m]: done

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m (hunk 14 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: done. Named them as MSID_BROKER_PROTOCOL_VERSION_2, MSID_BROKER_PROTOCOL_VERSION_3, MSID_BROKER_PROTOCOL_VERSION_4. Let me know if you have better names :)

File: IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m (hunk 13 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/token_request/MSIDBrokerOperationTokenRequest+Parameteres.m]: Agree, removed this todo.

File: IdentityCore/src/MSIDConstants.h [hunk 21]
```diff
@@ -86,7 +86,27 @@ extern NSString * _Nonnull const MSID_BROKER_APP_BUNDLE_ID;
...
+/**
+ This protocol is being used in ADAL for URL scheme requests to broker.
+ */
+extern NSInteger const MSID_BROKER_PROTOCOL_VERSION_2;
+/**
+This protocol is being used in MSAL for URL scheme requests to broker.
+*/
+extern NSInteger const MSID_BROKER_PROTOCOL_VERSION_3;
+/**
+This protocol is being used in MSAL for XPC requests to SSO extension.
```
Comment (oldalton) [inline IdentityCore/src/MSIDConstants.h]: will this be also used for MSAL in combination when SSO extension exists but we call broker through URL schemes? My assumption is that we would not return RT in that case, so I'm not clear what protocol version we should use there.

File: IdentityCore/src/broker_operation/request/MSIDBrokerOperationRequest.m [hunk 22]
```diff
@@ -40,17 +72,40 @@ - (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(NSError **)er
...
-    // TODO: implement.
+    NSMutableDictionary *json = [NSMutableDictionary new];
+    if (!self.brokerKey)
+    {
+        MSID_LOG_WITH_CORR(MSIDLogLevelError, self.correlationId, @"Failed to create json for %@ class, brokerKey is nil.", self.class);
+        return nil;
+    }
+    json[MSID_BROKER_KEY] = self.brokerKey;
+    json[MSID_BROKER_PROTOCOL_VERSION_KEY] = [@(self.protocolVersion) stringValue];
+    if (!self.protocolVersion) return nil;
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/MSIDBrokerOperationRequest.m]: nit: log?

File: IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest.m [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+    NSMutableDictionary *json = [[super jsonDictionary] mutableCopy];
+    if (!json) return nil;
+    
+    NSDictionary *accountIdentifierJson = [self.accountIdentifier jsonDictionary];
+    if (accountIdentifierJson) [json addEntriesFromDictionary:accountIdentifierJson];
+    
+    json[MSID_BROKER_LOGIN_HINT_KEY] = self.loginHint;
+    
+    NSString *promptString = MSIDPromptParamFromType(self.promptType);
+    if (!promptString) return nil;
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest.m]: nit: I'd recommend keeping symmetry in required parameters in serialization and deserialization. Here, valid prompt type seems mandatory, but not in initialization.

File: IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest.m [hunk 24]
```diff
@@ -46,17 +66,25 @@ - (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(NSError **)er
     
     if (self)
     {
-        // TODO: implement.
+        _accountIdentifier = [[MSIDAccountIdentifier alloc] initWithJSONDictionary:json error:error];
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest.m]: nit: add a log?

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m [hunk 25]
```diff
@@ -52,18 +53,54 @@ - (void)acquireToken:(MSIDRequestCompletionBlock)completionBlock
     {
         MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Interactive broker extension flow finished. Result %@, error: %ld error domain: %@", _PII_NULLIFY(result), (long)error.code, error.domain);
         
+        if ([self shouldFallback:error])
```
Comment (oldalton) [inline IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: nit: small optimization if you do if (error && [self shouldFallback:error]). That way you don't even call below method if there's no error and don't check fallbackController presence.

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+        __auto_type operationRequest = [MSIDBrokerOperationInteractiveTokenRequest tokenRequestWithParameters:self.requestParameters
+                                                                                                         error:&localError];
+        
+        if (!operationRequest)
+        {
+            completionBlock(nil, localError, nil);
+            return;
+        }
+        
+        ASAuthorizationSingleSignOnRequest *ssoRequest = [self.ssoProvider createRequest];
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: nit: does this need to be executed from main thread?

File: IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+
++ (ASAuthorizationSingleSignOnProvider *)msidSharedProvider
+{
+    static dispatch_once_t once;
+    static ASAuthorizationSingleSignOnProvider *ssoProvider;
+    
+    dispatch_once(&once, ^{
+        NSURL *url = [NSURL URLWithString:MSID_DEFAULT_AAD_AUTHORITY];
+    
+        ssoProvider = [ASAuthorizationSingleSignOnProvider authorizationProviderWithIdentityProviderURL:url];
```
Comment (oldalton) [inline IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m]: What if SSO extension presence status changes during process lifetime? Is ASAuthorizationSingleSignOnProvider smart enough in following scenario:

1. ASAuthorizationSingleSignOnProvider is created before SSO extension is activated, so canPerformAuthorization == NO
2. SSO extension got activated
3. Will same instance of ASAuthorizationSingleSignOnProvider now return canPerformAuthorization == YES?

Also, what are other pros and cons of using single shared instance vs creating new instance every time? I imagine, creating new instance should be relatively cheap. But if creation fails, we have a permanent failure for the lifetime of the process vs just for one request.

File: IdentityCore/src/webview/systemWebview/ios/MSIDAuthenticationSession.m [hunk 28]
```diff
@@ -38,7 +38,9 @@
 #import "MSIDBackgroundTaskManager.h"
 #if !MSID_EXCLUDE_WEBKIT
 #import <SafariServices/SafariServices.h>
-#import <AuthenticationServices/AuthenticationServices.h>
+    #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
```
Comment (oldalton) [inline IdentityCore/src/webview/systemWebview/ios/MSIDAuthenticationSession.m]: why was this added? This framework is available since iOS 12 and Xcode 10.

File: IdentityCore/src/broker_operation/request/MSIDBrokerOperationRequest.m (hunk 22 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/MSIDBrokerOperationRequest.m]: done

File: IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest.m (hunk 23 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/interactive_token_request/MSIDBrokerOperationInteractiveTokenRequest.m]: promptString will never be nil, removed this return

File: IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest.m (hunk 24 above)
Comment (antrix1989) [reply IdentityCore/src/broker_operation/request/silent_token_request/MSIDBrokerOperationSilentTokenRequest.m]: done

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m (hunk 26 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.m]: There is no such requirement.

File: IdentityCore/src/webview/systemWebview/ios/MSIDAuthenticationSession.m (hunk 28 above)
Comment (antrix1989) [reply IdentityCore/src/webview/systemWebview/ios/MSIDAuthenticationSession.m]: right, changed.

File: IdentityCore/src/MSIDConstants.h (hunk 21 above)
Comment (antrix1989) [reply IdentityCore/src/MSIDConstants.h]: We agreed to use current URL scheme protocol with extra flag. Will address this in separate pr.

File: IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m (hunk 25 above)
Comment (antrix1989) [reply IdentityCore/src/controllers/broker/MSIDSSOExtensionInteractiveTokenRequestController.m]: done

File: IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m (hunk 27 above)
Comment (antrix1989) [reply IdentityCore/src/util/ASAuthorizationSingleSignOnProvider+MSIDExtensions.m]: removed singleton
